File: frontend/src/types/detection.ts
export type RiskLevel = "SAFE" | "WARNING" | "DANGER";

export interface BoundingBox {
  x1: number;
  y1: number;
  x2: number;
  y2: number;
}

export interface DetectionItem {
  detection_id: string;
  source_model: string;
  class_id: number;
  class_name: string;
  confidence: number;
  bbox: BoundingBox;
  has_mask: boolean;
}

export interface PseudoBEVPoint {
  lateral: number;
  longitudinal: number;
}

export interface PseudoBEVRectangle {
  minimum_lateral: number;
  maximum_lateral: number;
  minimum_longitudinal: number;
  maximum_longitudinal: number;
}

export interface SafetyZones {
  footprint: PseudoBEVRectangle;
  danger: PseudoBEVRectangle;
  warning: PseudoBEVRectangle;
}

export interface PersonGeometry {
  person_id: string;
  confidence: number;
  bbox: BoundingBox;
  pseudo_bev_point: PseudoBEVPoint | null;
  mask_reliable: boolean;
  quality_reasons: string[];
}

export interface LoadGeometry {
  load_id: string;
  confidence: number;
  bbox: BoundingBox;
  pseudo_bev_points: PseudoBEVPoint[];
  safety_zones: SafetyZones | null;
  quality_reasons: string[];
}

export interface GeometryResponse {
  coordinate_system: "relative_pseudo_bev_not_metric";
  depth_low: number;
  depth_high: number;
  quality_reasons: string[];
  persons: PersonGeometry[];
  loads: LoadGeometry[];
}

export interface PairAssessment {
  person_id: string;
  load_id: string;
  risk_level: RiskLevel;
  matched_zone: RiskLevel | null;
  confidence: number;
  assessment_reliable: boolean;
  quality_reasons: string[];
}

export interface AssessmentResponse {
  risk_level: RiskLevel;
  assessment_reliable: boolean;
  quality_reasons: string[];
  contributing_person_ids: string[];
  contributing_load_ids: string[];
  pairs: PairAssessment[];
}

export interface DetectionSummary {
  person_count: number;
  load_count: number;
  rope_count: number;
}

export interface EvidenceResponse {
  rgb_url: string | null;
  pseudo_bev_url: string | null;
  combined_url: string | null;
}

export interface DepthMetadata {
  height: number;
  width: number;
  dtype: string;
  finite_min: number;
  finite_max: number;
  finite_fraction: number;
  convention: string;
}

export interface ProcessingMetadata {
  pipeline_version: string;
  frame_id: string;
  image_width: number;
  image_height: number;
  depth: DepthMetadata;
  models_loaded: Record<string, boolean>;
  config_versions: Record<string, string>;
}

export interface ImageDetectionResponse {
  status: "completed";
  processing_time_ms: number;
  assessment: AssessmentResponse;
  summary: DetectionSummary;
  detections: DetectionItem[];
  geometry: GeometryResponse;
  evidence: EvidenceResponse;
  metadata: ProcessingMetadata;
}

export type VideoJobStatus = "queued" | "processing" | "completed" | "failed";

export interface VideoJobCreated {
  job_id: string;
  status: "queued";
  status_url: string;
  stream_url: string;
  frame_results_url: string;
  result_url: string;
}

export interface VideoFrameRiskResult {
  frame_number: number;
  timestamp_seconds: number;
  risk_level: RiskLevel;
}

export interface VideoFrameRiskResultsPage {
  job_id: string;
  job_status: VideoJobStatus;
  items: VideoFrameRiskResult[];
  next_after_frame: number;
  has_more: boolean;
}

export interface VideoSummary {
  processed_frames: number;
  safe_frames: number;
  warning_frames: number;
  danger_frames: number;
  max_risk_level: RiskLevel | null;
  average_processing_fps: number;
  risk_segment_count: number;
}

export interface RiskSegment {
  segment_id: string;
  start_frame: number;
  end_frame: number;
  risk_start_frame: number;
  risk_end_frame: number;
  start_seconds: number;
  end_seconds: number;
  max_risk_level: "WARNING" | "DANGER";
  warning_frame_count: number;
  danger_frame_count: number;
  frame_evidence: VideoFrameEvidence[];
  result_url: string;
  output_codec: string;
  browser_playback_compatible: boolean;
  playback_warning: string | null;
}

export interface VideoFrameEvidence {
  frame_number: number;
  timestamp_seconds: number;
  risk_level: "WARNING" | "DANGER";
  original_url: string;
  rgb_url: string;
  pseudo_bev_url: string;
}

export interface VideoJob {
  job_id: string;
  status: VideoJobStatus;
  input_path: string;
  output_path: string;
  current_frame: number;
  total_frames: number;
  progress: number;
  processing_fps: number;
  elapsed_seconds: number;
  current_risk_level: RiskLevel | null;
  max_risk_level: RiskLevel | null;
  safe_frame_count: number;
  warning_frame_count: number;
  danger_frame_count: number;
  error: string | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
  stream_url: string;
  frame_results_url: string;
  result_url: string | null;
  download_url: string | null;
  report_url: string | null;
  summary: VideoSummary | null;
  risk_segments: RiskSegment[];
  output_codec: string;
  browser_playback_compatible: boolean;
  playback_warning: string | null;
}

export interface VideoReportEvidence {
  frame_number: number;
  timestamp_seconds: number;
  risk_level: "WARNING" | "DANGER";
  original_url: string;
  rgb_url: string;
  pseudo_bev_url: string;
}

export interface VideoReportSegment {
  segment_id: string;
  start_frame: number;
  end_frame: number;
  risk_start_frame: number;
  risk_end_frame: number;
  start_seconds: number;
  end_seconds: number;
  max_risk_level: "WARNING" | "DANGER";
  warning_frame_count: number;
  danger_frame_count: number;
  result_url: string;
  codec: string;
  browser_playback_compatible: boolean;
  playback_warning: string | null;
  evidence: VideoReportEvidence[];
}

export interface VideoReport {
  schema_version: string;
  job_id: string;
  status: "completed";
  created_at: string;
  started_at: string | null;
  completed_at: string;
  input_filename: string;
  summary: {
    processed_frames: number;
    total_frames: number;
    safe_frames: number;
    warning_frames: number;
    danger_frames: number;
    max_risk_level: RiskLevel | null;
    average_processing_fps: number;
    elapsed_seconds: number;
    risk_segment_count: number;
  };
  video: {
    filename: string;
    url: string;
    download_url: string;
    codec: string;
    browser_playback_compatible: boolean;
    playback_warning: string | null;
  };
  risk_segments: VideoReportSegment[];
}

File: backend/app/schemas/detection.py
"""Pydantic response contracts for image safety assessment."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

RiskLevelValue = Literal["SAFE", "WARNING", "DANGER"]


class StrictSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")


class BoundingBox(StrictSchema):
    x1: float
    y1: float
    x2: float
    y2: float


class DetectionItem(StrictSchema):
    detection_id: str
    source_model: str
    class_id: int
    class_name: str
    confidence: float = Field(ge=0.0, le=1.0)
    bbox: BoundingBox
    has_mask: bool


class PseudoBEVPoint(StrictSchema):
    lateral: float
    longitudinal: float


class PseudoBEVRectangle(StrictSchema):
    minimum_lateral: float
    maximum_lateral: float
    minimum_longitudinal: float
    maximum_longitudinal: float


class SafetyZones(StrictSchema):
    footprint: PseudoBEVRectangle
    danger: PseudoBEVRectangle
    warning: PseudoBEVRectangle


class PersonGeometry(StrictSchema):
    person_id: str
    confidence: float
    bbox: BoundingBox
    pseudo_bev_point: PseudoBEVPoint | None
    mask_reliable: bool
    quality_reasons: list[str]


class LoadGeometry(StrictSchema):
    load_id: str
    confidence: float
    bbox: BoundingBox
    pseudo_bev_points: list[PseudoBEVPoint]
    safety_zones: SafetyZones | None
    quality_reasons: list[str]


class GeometryResponse(StrictSchema):
    coordinate_system: Literal["relative_pseudo_bev_not_metric"] = (
        "relative_pseudo_bev_not_metric"
    )
    depth_low: float
    depth_high: float
    quality_reasons: list[str]
    persons: list[PersonGeometry]
    loads: list[LoadGeometry]


class PairAssessment(StrictSchema):
    person_id: str
    load_id: str
    risk_level: RiskLevelValue
    matched_zone: RiskLevelValue | None
    confidence: float = Field(ge=0.0, le=1.0)
    assessment_reliable: bool
    quality_reasons: list[str]


class AssessmentResponse(StrictSchema):
    risk_level: RiskLevelValue
    assessment_reliable: bool
    quality_reasons: list[str]
    contributing_person_ids: list[str]
    contributing_load_ids: list[str]
    pairs: list[PairAssessment]


class DetectionSummary(StrictSchema):
    person_count: int = Field(ge=0)
    load_count: int = Field(ge=0)
    rope_count: int = Field(ge=0)


class EvidenceResponse(StrictSchema):
    rgb_url: str | None = None
    pseudo_bev_url: str | None = None
    combined_url: str | None = None


class DepthMetadata(StrictSchema):
    height: int
    width: int
    dtype: str
    finite_min: float
    finite_max: float
    finite_fraction: float
    convention: str


class ProcessingMetadata(StrictSchema):
    pipeline_version: str
    frame_id: str
    image_width: int
    image_height: int
    depth: DepthMetadata
    models_loaded: dict[str, bool]
    config_versions: dict[str, str]


class ImageDetectionResponse(StrictSchema):
    status: Literal["completed"] = "completed"
    processing_time_ms: float = Field(ge=0.0)
    assessment: AssessmentResponse
    summary: DetectionSummary
    detections: list[DetectionItem]
    geometry: GeometryResponse
    evidence: EvidenceResponse
    metadata: ProcessingMetadata


__all__ = ["ImageDetectionResponse", "RiskLevelValue"]
